### PR TITLE
Add Photon notifications to AMO

### PR DESCRIPTION
**Pull request by kumar303**
Fixes https://github.com/mozilla/addons-frontend/issues/3706

I needed success notices for the add-to-collection feature so I implemented most of [the Photon notifications](https://design.firefox.com/photon/components/message-bars.html). 

These features have *not* been implemented:
* Generic notices or warnings
* Discopane. I started trying to update this code but it was copy/pasted and has complications due to the flex rules. I can't fit it into this patch.
* Messages are not yet dismissable

screenshot 2018-01-16 11 33 27
screenshot 2018-01-12 12 34 00
screenshot 2018-01-16 11 34 29

Here is a success notice:

screenshot 2018-01-12 08 52 50

Here's an error on the review page:

screenshot 2018-01-16 11 29 51

Here's how it looks with an action button:

screenshot 2018-01-16 11 25 51
screenshot 2018-01-12 09 59 16
screenshot 2018-01-12 09 59 35
screenshot 2018-01-16 11 27 14
screenshot 2018-01-12 10 01 11

### CODE REVIEW

**Comment by willdurand** [comment]
😍

**Comment by tofumatt** [comment]
It looks great! 👍 I'll probably get to the review later tonight or tomorrow, but thanks for this!

File: src/ui/components/Notice/index.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
+/* @flow */
+import makeClassName from 'classnames';
+import * as React from 'react';
```
Comment (tofumatt) [inline src/ui/components/Notice/index.js]: Why not just `import React from 'react';`?

File: src/ui/components/Notice/styles.scss [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+@import "~ui/css/vars";
+@import "~photon-colors/colors";
+
+.Notice {
+  border-radius: $border-radius-s;
+  display: grid;
+  font-size: $font-size-s;
+  grid-gap: 4px;
+  grid-template-columns: 20px auto;
+  line-height: 15px;
```
Comment (tofumatt) [inline src/ui/components/Notice/styles.scss]: `line-height` [should be in unitless measurements](https://css-tricks.com/almanac/properties/l/line-height/#article-header-id-0); I'm guessing this is probably around `1.2` but I'm not sure, you'll have to trial+error it 🙃

File: src/ui/components/Notice/styles.scss [hunk 3]
```diff
@@ -0,0 +1,97 @@
+@import "~core/css/inc/mixins";
+@import "~core/css/inc/vars";
+@import "~ui/css/vars";
+@import "~photon-colors/colors";
+
+.Notice {
+  border-radius: $border-radius-s;
+  display: grid;
+  font-size: $font-size-s;
+  grid-gap: 4px;
```
Comment (tofumatt) [inline src/ui/components/Notice/styles.scss]: Most of our base units are `6px` multiples. Should this really be `4px`? If it totally looks right compared to mocks that's okay, but I'd assume 6 is preferred.

File: src/ui/components/Notice/styles.scss [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+  grid-template-columns: 20px auto;
+  line-height: 15px;
+  min-height: 20px;
```
Comment (tofumatt) [inline src/ui/components/Notice/styles.scss]: Same throughout re: `6px` base multiple; these all seem to be multiples of `4` which isn't common in Add-ons or Photon and is in fact a new multiple.

File: src/ui/components/Notice/styles.scss [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+  // end of the text block when there's no button.
+  @include padding-end(4px);
+}
+
+@mixin notice_button() {
+  &,
+  &.Button--small {
+    border: none;
+    font-size: $font-size-s;
+    line-height: 15px;
```
Comment (tofumatt) [inline src/ui/components/Notice/styles.scss]: This should be unitless too.

File: src/ui/components/Notice/styles.scss [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    padding: 6px;
+    text-shadow: none;
+    width: max-content;
```
Comment (tofumatt) [inline src/ui/components/Notice/styles.scss]: This seems to be unsupported in Edge: https://caniuse.com/#search=max-content

That's not the end of the world IMO but I'd rather us be compatible. Is there any way to replicate what you're doing here and have it work with Edge?

File: tests/unit/ui/components/TestNotice.js [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+  it('requires a known type', () => {
+    expect(() => render({ type: 'nope' })).toThrow(/Unknown type: nope/);
+  });
+
+  it('hides action buttons by default', () => {
+    const root = render();
+
+    expect(root).not.toHaveClassName('.Notice-with-action');
+    // There should only be one column besides the icon.
```
Comment (tofumatt) [inline tests/unit/ui/components/TestNotice.js]: This is an uber-nitpick but "besides the icon" is a bit confusing to read–I couldn't tell at first if it meant "other than the icon" or "next to the icon". I think "one column other than the icon column" would clarify things.

**Comment by kumar303** [comment]
Thanks for the review! I'm back from the holiday so I'll start working on this after breakfast. Did you look at [the spec](https://design.firefox.com/photon/components/message-bars.html)? It's pretty nice because it has pixel dimensions and it specifies button placement, etc. Some changes you requested go against that spec. If you would like to do that, we should probably get approval from someone on the Photon project.

File: src/ui/components/Notice/index.js (hunk 1 above)
Comment (kumar303) [reply src/ui/components/Notice/index.js]: The [package](https://reactjs.org/blog/2017/09/26/react-v16.0.html#packaging) changed in React 16 so importing `*` makes more things work as expected.

File: src/ui/components/Notice/styles.scss (hunk 3 above)
Comment (kumar303) [reply src/ui/components/Notice/styles.scss]: The [spec](https://design.firefox.com/photon/components/message-bars.html) says it should be 4px.

**Comment by tofumatt** [comment]
Oh, I see. I hadn’t seen that part of the spec.

I definitely don’t like that 😅 But you’re right: it’s part of the spec. I’ll bring it up with people but you’re right: you should follow the spec for now.

File: src/ui/components/Notice/styles.scss (hunk 6 above)
Comment (tofumatt) [reply src/ui/components/Notice/styles.scss]: Ah, okay, that's fine then. 👍

**Comment by kumar303** [comment]
I updated the multi-line message based on a "spec" that @pwalm just gave me 😃 I think this looks better (I adjusted the screenshots). 
img_1255

**Comment by kumar303** [comment]
Tests + lint all pass for me locally

**Pull request merged**